docs/advanced-usage.rst

```diff
@@ -0,0 +1,176 @@
+Advanced Usage
+==============
+
+.. currentmodule:: urllib3
+
+.. _stream:
+
+Streaming and IO
+----------------
+
+When dealing with large responses it's often better to stream the response
+content::
+
+    >>> r = http.request(
+    ...     'GET',
+    ...     'http://httpbin.org/bytes/1024',
+    ...     preload_content=False)
+    >>> for chunk in r.stream(32):
+    ...     print(chunk)
+    b'...'
+    b'...'
+    ...
+    >>> r.release_conn()
+
+Setting ``preload_content`` to ``False`` means that urllib3 will stream the
+response content. :meth:`~response.HTTPResponse.stream` lets you iterate over
+chunks of the response content.
+
+.. note:: When using ``preload_content=False``, you should call 
+    :meth:`~response.HTTPResponse.release_conn` to release the http connection
+    back to the connection pool so that it can be re-used.
+
+However, you can also treat the :class:`~response.HTTPResponse` instance as
+a file-like object. This allows you to do buffering::
+
+    >>> r = http.request(
+    ...     'GET',
+    ...     'http://httpbin.org/bytes/1024',
+    ...     preload_content=False)
+    >>> r.read(4)
+    b'\x88\x1f\x8b\xe5'
+
+Calls to :meth:`~response.HTTPResponse.read()` will block until more response
+data is available. 
+
+    >>> import io
+    >>> reader = io.BufferedReader(r, 8)
+    >>> reader.read(4)
+    >>> r.release_conn()
+
+You can use this file-like object to do things like decode the content using
+:mod:`codecs`::
+
+    >>> import codecs
+    >>> reader = codecs.getreader('utf-8')
+    >>> r = http.request(
+    ...     'GET',
+    ...     'http://httpbin.org/ip',
+    ...     preload_content=False)
+    >>> json.load(reader(r))
+    {'origin': '127.0.0.1'}
+    >>> r.release_conn()
+
+.. _proxies:
+
+Proxies
+-------
+
+You can use :class:`~poolmanager.ProxyManager` to tunnel requests through an
+HTTP proxy::
+
+    >>> proxy = urllib3.ProxyManager('http://localhost:3128/')
+    >>> proxy.request('GET', 'http://google.com/')
+
+The usage of :class:`~poolmanager.ProxyManager` is the same as
+:class:`~poolmanager.PoolManager`.
+
+You can use :class:`~contrib.socks.SOCKSProxyManager` to connect to SOCKS4 or
+SOCKS5 proxies. In order to use SOCKS proxies you will need to install
+`PySocks <https://pypi.python.org/pypi/PySocks>`_ or install urllib3 with the
+``socks`` extra::
+
+    pip install urllib3[socks]
+
+Once PySocks is installed, you can use
+:class:`~contrib.socks.SOCKSProxyManager`::
+
+    >>> from urllib3.contrib.socks import SOCKSProxyManager
+    >>> proxy = SOCKSProxyManager('socks5://localhost:8889/')
+    >>> proxy.request('GET', 'http://google.com/')
+
+.. _ssl_mac:
+
+Certificate validation and Mac OS X
+-----------------------------------
+
+Apple-provided Python and OpenSSL libraries contain a patches that make them
```

Comment:
This could possibly use some clarification; if we're going to bring it up, it'd be nice to know the exact cases where it's going to be a problem, along with a recipe to avoid it (from what I can tell, `brew install openssl; brew install python --with-brewed-openssl`).